default/AI/ResourcesAI.py

```diff
@@ -29,98 +38,258 @@
+
+
+    def bake_future_focus(self, pid, focus, update=True):
+        """Set the focus and moves it from the raw list to the baked list of planets
+        Return success or failure
+        """
+        pinfo = self.raw_planet_info.get(pid)
+        success = (pinfo is not None and
+                   (pinfo.current_focus == focus
+                    or (focus in pinfo.planet.availableFoci
+                        and fo.issueChangeFocusOrder(pid, focus))))
+        if success:
+            pinfo.future_focus = focus
+            self.baked_planet_info[pid] = self.raw_planet_info.pop(pid)
+
+            if update and pinfo.current_focus != focus:
+                # TODO determine if this should update all planets (for reporting) or just the remaining unplaced planets
+                universe = fo.getUniverse()
+                universe.updateMeterEstimates(self.raw_planet_info.keys())
+                itarget = pinfo.planet.currentMeterValue(fo.meterType.targetIndustry)
+                rtarget = pinfo.planet.currentMeterValue(fo.meterType.targetResearch)
+                pinfo.possible_output[focus] = (itarget, rtarget)
+        return success
+
+
+    def calculate_planet_infos(self, pids):
+        """ Calculates for each possible focus the target output of each planet and stores it in planet info
+        It excludes baked planets from consideration.
+        Note: The results will not be strictly correct if any planets have global effects
+        """
+        #TODO this function depends on the specific rule that off-focus meter value are always the minimum value
+        universe = fo.getUniverse()
+        unbaked_pids = [pid for pid in pids if not pid in self.baked_planet_info]
+        planet_infos = [(pid, self.all_planet_info[pid], self.all_planet_info[pid].planet) for pid in unbaked_pids]
+        for pid, pinfo, planet in planet_infos:
+            if IFocus in planet.availableFoci and planet.focus != IFocus:
+                fo.issueChangeFocusOrder(pid, IFocus)  # may not be able to take, but try
+
+        universe.updateMeterEstimates(unbaked_pids)
+        for pid, pinfo, planet in planet_infos:
+            itarget = planet.currentMeterValue(fo.meterType.targetIndustry)
+            rtarget = planet.currentMeterValue(fo.meterType.targetResearch)
+            if planet.focus == IFocus:
+                pinfo.possible_output[IFocus] = (itarget, rtarget)
+                pinfo.possible_output[GFocus] = rtarget
+            else:
+                pinfo.possible_output[IFocus] = (0, 0)
+                pinfo.possible_output[GFocus] = 0
+            if RFocus in planet.availableFoci and planet.focus != RFocus:
+                fo.issueChangeFocusOrder(pid, RFocus)  # may not be able to take, but try
+
+        universe.updateMeterEstimates(unbaked_pids)
+        for pid, pinfo, planet in planet_infos:
+            can_focus = planet.currentMeterValue(fo.meterType.targetPopulation) > 0
```

Comment:
That is not the only place this check is needed. Maybe define it as function in ```PlanetFocusInfo```?